// modOptions/options.js
/**
 * Kili.us server options module
 *
 * Author: Jeff Horak
 * Date: 6/6/12
 */

var optionPath = '../../../',
    options = require(optionPath + 'options.json'),
    fs = require('fs'),

    defaults = {
      database: 'kilius',
      serverPort: 8642,
      databaseHost: 'localhost',
      databasePort: 27017,
      databaseUser: '',
      databasePassword: '',
      throttleLimit: 25,
      throttleTime: 86400000,
      statsPageLimit: 10,
      clientRootFilePath: '../Kilius-client/src'
    };

/**
 * Populate the exported object with the property or the default
 */
(function() {
  var clientFilePath;

  for(var prop in defaults) {
    if (defaults.hasOwnProperty(prop)) {

      //properties.push(prop);
      // Set the defaults on the exported object
      exports[prop] = options[prop] || defaults[prop];
    }
  }

  clientFilePath = exports.clientRootFilePath;

  if (clientFilePath[0] !== '/') {
    // This is not an absolute path, calculate the absolute path
    //   Do it synchronously, this is part of the application start-up
    exports.clientRootFilePath = fs.realpathSync(__dirname + optionPath + '../' + clientFilePath);
  }

})();
