// modLogging/logging.js
/**
 * Fire and forget logging module for Kili.us server
 *
 * Author: Jeff Horak
 * Date: 7/6/12
 */

var db = require('modDatabase'),
    serviceMode = false;

exports.setServiceMode = function(isService) {
  serviceMode = isService;
};

exports.isServiceMode = function() {
  return serviceMode;
};

// Write to the activity log
exports.log = function(payload) {

  var message = payload ? payload.message : null,
      now = new Date();

  if (message && serviceMode) {
    console.log( "[" + now.toLocaleString() + "] " + message);
  }

  db.logActivity(payload).then(
      // Logged successfully, nothing left to do
      null,

      // There was an error logging the activity
      function(err) {

        // Need to write to the error log
        exports.error({
          message: 'Database error writing to the activity log',
          error: err.error,
          code: err.code
        });
      }
  );
};

exports.error = function(payload) {

  var message = payload ? payload.message : null,
      red = '\u001b[31m',
      reset = '\u001b[0m',
      now = new Date();

  if (message) {
    console.error( "[" + now.toLocaleString() + "] " +
                   (serviceMode ? '' : red) +
                   message +
                   (serviceMode ? '' : reset) );
  }

  db.logError(payload).then(
      // Logged successfully, nothing left to do
      null,

      // There was an error writing to the error log, throw
      function(err) {
        throw err;
      }
  );
};
