// modStats/statService.js
/**
 * Statistics service for Kili.us server
 * Author: Jeff Horak
 * Date: 7/6/12
 */

var db = require('modDatabase'),
    logging = require('modLogging'),
    Promise = require('modPromise').Promise,
    pageSize = require('modOptions').statsPageLimit;

exports.linksForUser = function(options) {
  var promise = new Promise(),
      client = options.clientID;

  db.linksForUser({
    page: options.page,
    pageSize: pageSize,
    clientID: client
  }).then(

      // Fetched the requested links for the user
      function(results) {

        promise.resolve(results);

        // Success - log activity
        logging.log({
          message: 'Fetched links for user',
          client: client
        });

      },

      // Some error returned from the database
      function(err) {
        err = err || {};

        if (!err.message) {
          err.message = 'Database error';
        }

        if (!err.code) {
          err.code = 500;
        }

        // Reject the promise
        promise.reject(err.message, err.code);

        // Log the error
        logging.error(err);
      }
  );

  return promise;
};
