// modStats/statService.js
//////////////////////////////////////////////////////////////////////////
// File: statService.js
//
// Description: 
//   Statistics service for kili.us
//
//////////////////////////////////////////////////////////////////////////

exports.linksForUser = function(data, cb, errCB) {
  var links = data.database.collection('links'),
      pageSize = 10,
      skipTo = data.page || 0,
      cursor = links.find({clientID: data.clientID}, {'longLink': 1, 'shortLink': 1, 'hits': 1})
                    .limit(pageSize, null)
                    .skip(pageSize*skipTo, null),
      stat = [];

  cursor.each(function(err, doc) {
    if (err) {
      errCB({
        message: 'Oops. We\'ve encountered a database error and were unable to process your request.',
        err: err,
        log: 'Error building stats for user ' + data.clientID
      });
    } else if (doc === null) {
      // That's the last item, return the data
      cb(stat);
    } else {
      // Push onto the stat array
      stat.push({
        long: doc.longLink,
        short: doc.shortLink,
        hits: doc.hits.length || 0
      });
    }
  });
};
